README.md
# which_plates

CI/CD Status:  
![Release](https://img.shields.io/github/v/release/iandday/whichPlates?color=brightgreen&sort=semver)
[![Linting](https://github.com/iandday/whichPlates/actions/workflows/lint.yml/badge.svg)](https://github.com/iandday/whichPlates/actions/workflows/lint.yml)
[![Testing](https://github.com/iandday/whichPlates/actions/workflows/test.yml/badge.svg)](https://github.com/iandday/whichPlates/actions/workflows/test.yml)
[![Documentation](https://github.com/iandday/whichPlates/actions/workflows/doucmentation.yml/badge.svg)](https://github.com/iandday/whichPlates/actions/workflows/documentation.yml)

<!-- Pytest Coverage Comment:Begin -->
<!-- Pytest Coverage Comment:End -->

Calculate weight and plates needed for each set based on one rep max weight

```code=bash
ian@Ians-MacBook-Pro whichPlates % which-plates
Bar Weight:45
One rep max:250
Percentages for each set, space separated: 50 65
Set 1 @ 50%: 125.0lbs, 80.0lbs in plates
     35: 2
     5: 2
Set 2 @ 65%: 165.0lbs, 120.0lbs in plates
     45: 2
     15: 2
Plates Needed:
     45: 2
     35: 2
     15: 2
     5: 2
```


CHANGELOG.md
# Changelog

<!--next-version-placeholder-->

## v0.9.0 (2023-05-01)
### Feature
* Fix rounding ([`3198e9b`](https://github.com/iandday/whichPlates/commit/3198e9b88b3b59979556dc6088ec9f539b4041cd))

## v0.8.0 (2023-04-30)
### Feature
* Fixed math issue ([`364cb3c`](https://github.com/iandday/whichPlates/commit/364cb3ccdd5be0ba9837bf1ccb749aff79d9a2e9))

### Fix
* Linting ([`13e45bc`](https://github.com/iandday/whichPlates/commit/13e45bc61d96b53374e0d0e2584cf8a59fdc1d7f))

## v0.7.0 (2023-04-30)
### Feature
* :bug: Sort list of plates first ([`c7168b1`](https://github.com/iandday/whichPlates/commit/c7168b15a1ab3a2f5f46a1a5b92868074356980b))

## v0.6.0 (2023-04-29)
### Feature
* Typo ([`afe7601`](https://github.com/iandday/whichPlates/commit/afe760187eec2168006d808344ef9c00c8ddf5b6))
* **functions:** Actions ([`aa86de3`](https://github.com/iandday/whichPlates/commit/aa86de32b5280a578cb625f7a39309b89fc9e552))

## v0.5.0 (2023-04-29)
### Feature
* Workflows ([`230b001`](https://github.com/iandday/whichPlates/commit/230b001a0eea476e060adf17e4fe7d4d9858571c))

## v0.4.0 (2023-04-29)
### Feature
* Updated gitignore ([`50b78fa`](https://github.com/iandday/whichPlates/commit/50b78fa1179a9c13e5965236fd7b5a93f1611d6e))

## v0.3.0 (2023-04-29)
### Feature
* **functions:** Update actions ([`04ccc3e`](https://github.com/iandday/whichPlates/commit/04ccc3e4fe6702aab5a7daf95f9c12a66a538f4c))
* **functions:** Update actions ([`ca831f4`](https://github.com/iandday/whichPlates/commit/ca831f445e212affd998f9b29c1ee8ab81823d60))

## v0.2.0 (2023-04-19)
### Feature
* **functions:** Update docstrings ([`fef4d29`](https://github.com/iandday/whichPlates/commit/fef4d29f0c7f0cab9077f60ae218f6215a3d9a13))


tests/test_main.py
"""test main.py"""

import pytest

from which_plates.main import main, main_function


def test_main(monkeypatch) -> None:  # type: ignore[no-untyped-def]
    """test main"""

    # fake inputs
    bar_weight = 45
    total_weight = 245
    percentages = "50 75 95"
    answers = iter([bar_weight, total_weight, percentages])
    monkeypatch.setattr("builtins.input", lambda name: next(answers))

    with pytest.raises(SystemExit) as excinfo:
        main()
    assert excinfo.value.code == 0


def test_main_function() -> None:
    """test main_function"""

    bar_weight = 45
    total_weight = 245
    percentages = [0.50, 0.75, 0.95]
    available_plates = [45, 35, 25, 15, 10, 5, 2.5]

    result = main_function(bar_weight, total_weight, available_plates, percentages)
    assert result == 0


tests/test_functions.py
"""which_plates tests"""

import pytest

from which_plates.functions import calc_plates, round_num


def test_round_num_float() -> None:
    """Float input"""
    result = round_num(16.5)
    assert result == 20


def test_round_num_int() -> None:
    """Integer input"""
    result = round_num(16)
    assert result == 20


def test_round_num_string() -> None:
    """String input"""

    with pytest.raises(SystemExit) as pytest_wrapped_e:
        round_num("g")  # type: ignore[arg-type]
    assert pytest_wrapped_e.type == SystemExit
    assert pytest_wrapped_e.value.code == 1


def test_calc_plates_int_and_list() -> None:
    """expected input"""
    result = calc_plates(200, [45, 35, 25, 15, 10, 5, 2.5])
    assert isinstance(result, dict)
    assert result == {45: 4, 10: 2}


src/which_plates/tests/test_main.py
'''test main.py'''

import pytest

from which_plates.main import  main, main_function


def test_main(monkeypatch):
    ''' test main'''

    # fake inputs
    bar_weight = 45
    total_weight = 245
    percentages = '50 75 95'
    answers = iter([bar_weight, total_weight, percentages])
    monkeypatch.setattr('builtins.input', lambda name: next(answers))

    with pytest.raises(SystemExit) as excinfo:
        main()
    assert excinfo.value.code ==0


def test_main_function():
    '''test main_function'''

    bar_weight = 45
    total_weight = 245
    percentages = [.50, .75, .95]
    available_plates = [45, 35, 25, 15, 10, 5, 2.5]

    result = main_function(bar_weight, total_weight, available_plates, percentages)
    assert result == 0


src/which_plates/main.py
""" main function for which_plates"""

import sys
from . import functions


def main_function(bar_weight, rep_max, available_plates, percentages):
    """main function
    """

    set_count = 1
    used_plates = {}

    for plates in available_plates:
        used_plates[plates] = 0

    for percent in percentages:

        # calculate plate weight needed for current percentage
        plate_weight = functions.round_num((rep_max * percent) - bar_weight)

        if plate_weight <= 0:
            print('    Computed weight less than bar weight, try again')
        else:
            total_weight = round(plate_weight + bar_weight)
            weight = round(plate_weight)
            print(F"Set {set_count} @ {percent * 100}%: {total_weight}lbs, {weight}lbs in plates")  # noqa: E501

        plates = functions.calc_plates(plate_weight, available_plates)
        for p_weight, p_count in plates.items():
            if used_plates[p_weight] < p_count:
                used_plates[p_weight] = p_count
        set_count += 1

    # Print summary
    for p_weight, p_count in used_plates.items():
        if p_count:
            print(F'     {p_weight}: {p_count}')

    return 0


def main():
    """function with no inputs to allow use as a console script"""

    available_plates = [45, 35, 25, 15, 10, 5, 2.5]
    bar_weight = int(input("Bar Weight:"))
    rep_max = int(input("One rep max:"))
    raw_list = input("Percentages for each set, space separated: ").split()
    percentages = [float(n)/100 for n in raw_list if n.isdigit()]

    sys.exit(main_function(bar_weight, rep_max, available_plates, percentages))


if __name__ == "__main__":

    main()


src/which_plates/functions.py
"""Supporting functionss"""

from decimal import Decimal


def round_num(number: float) -> int:
    """rounds to nearest multiple of 5, assuming smallest plate is 2.5 pounds
    Args:
        number (float): number to be rounded
    Returns:
        int: rounded number
    """

    try:
        number = float(number)
    except ValueError as value_error:
        raise SystemExit(1) from value_error

    if number % 10 < 5:
        rounded_number = number - (number % 10)
    elif number % 10 > 5:
        rounded_number = number + (10 - (number % 10))
    else:
        rounded_number = number
    return int(rounded_number)


def calc_plates(weight: int, available_plates: list[float]) -> dict:
    """computes the required weight plates to reach supplied weight
    Args:
        weight (int): desired weight
        available_plates (list): available plates
    Returns:
        dict: number of plates per weight to reach supplied weight
    """
    plates = {}
    remaining_weight = Decimal(round_num(weight))
    available_plates.sort(reverse=True)

    for plate in available_plates:
        # at least two plates left in remaining weight
        if remaining_weight / Decimal(plate) >= Decimal(2):

            # can only use even number of plates
            if ((remaining_weight / Decimal(plate)) %2)==0:
                plate_count = int(Decimal(remaining_weight) // Decimal(plate))
            else:
                plate_count = int(Decimal(remaining_weight) // Decimal(plate))-1
            plates[plate] = plate_count

            remaining_weight -= Decimal(plate) * Decimal(plates[plate])

    return plates


whichPlates.py
#!/usr/bin/env python

"""whichPlates.py: Returns weight and plates needed per set based on 1 RM."""

from decimal import *


##############################################################
# func_input_sets()
# This function prompts user for percentages of each set and
# returns list of percentages
# inputs: None
# returns: List of 1 rep max percentages.
def func_input_sets():
    list = raw_input("Percentages for each set, space separated: ").split()
    return [float(n)/100 for n in list if n.isdigit()]


##############################################################



##############################################################



if __name__ == "__main__":
    setCount = 1
    weightInSet = []
    usedPlates = {}
    availablePlates = [45, 35, 25, 15, 10, 5, 2.5]


    barWeight = int(input("Bar Weight:"))
    repMax = int(input("One rep max:"))
    workingWeight = repMax - barWeight
    percentages = func_input_sets()

    for plates in availablePlates:
        usedPlates[plates] = 0

    for percent in percentages:
        weight = func_round_num((repMax * percent) - barWeight)
        if weight <= 0:
            print '    Computed weight less than bar weight, try again'
        else:
            print "Set " + str(setCount) + " @ " + str(percent * 100) + '%: ' + str((weight + barWeight)) + "lbs, " + str(weight) + "lbs in plates"
        plates = func_calc_plates(weight, availablePlates)

        for plate in plates:
            if usedPlates[plate] < plates[plate]:
                usedPlates[plate] = plates[plate] * 2
            print "     " + str(plate) + ":" + str(plates[plate] * 2)
        setCount += 1

    # Print summary
    print 'Plates Needed:'
    for plate in usedPlates:
        if usedPlates[plate]:
            print '     ' + str(plate) + ': ' + str(usedPlates[plate])
